internal/cache/file/downloader/job.go

```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
+// notify the subscribers.
+func (job *Job) updateStatusOffset(downloadedOffset int64) (err error) {
 	fileInfoKey := data.FileInfoKey{
 		BucketName: job.bucket.Name(),
 		ObjectName: job.object.Name,
 	}
 	fileInfoKeyName, err := fileInfoKey.Key()
 	if err != nil {
-		err = fmt.Errorf("updateFileInfoCache: error while creating fileInfoKeyName for bucket %s and object %s %w",
+		err = fmt.Errorf("updateStatusOffset: error while creating fileInfoKeyName for bucket %s and object %s %w",
 			fileInfoKey.BucketName, fileInfoKey.ObjectName, err)
 		return
 	}
 
 	updatedFileInfo := data.FileInfo{
 		Key: fileInfoKey, ObjectGeneration: job.object.Generation,
-		FileSize: job.object.Size, Offset: uint64(job.status.Offset),
+		FileSize: job.object.Size, Offset: uint64(downloadedOffset),
 	}
-
-	logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+	job.mu.Lock()
+	defer job.mu.Unlock()
 	err = job.fileInfoCache.UpdateWithoutChangingOrder(fileInfoKeyName, updatedFileInfo)
-	if err != nil {
-		err = fmt.Errorf("updateFileInfoCache: error while inserting into fileInfoCache %s: %w", updatedFileInfo.Key, err)
-		return
+	if err == nil {
+		job.status.Offset = downloadedOffset
+		// Notify subscribers if file cache is updated.
+		logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+		job.notifySubscribers()
+	} else {
+		err = fmt.Errorf("updateStatusOffset: error while updating offset: %v in fileInfoCache %s: %w", downloadedOffset, updatedFileInfo.Key, err)
 	}
 	return
 }
 
+// downloadObjectToFile downloads the backing object from GCS into the given
+// file and updates the file info cache. It uses gcs.Bucket's NewReader method
+// to download the object.
+func (job *Job) downloadObjectToFile(cacheFile *os.File) (err error) {
+	var newReader io.ReadCloser
+	var start, end, sequentialReadSize, newReaderLimit int64
+	end = int64(job.object.Size)
+	sequentialReadSize = int64(job.sequentialReadSizeMb) * cacheutil.MiB
+
+	// Each iteration of this for loop, reads ReadChunkSize size of range of the
+	// backing object from reader into the file handle and updates the file info
+	// cache. In case, reader is not present for reading, it creates a
+	// gcs.Bucket's NewReader with size min(sequentialReadSize, object.Size).
+	for start < end {
+		if newReader == nil {
+			newReaderLimit = min(start+sequentialReadSize, end)
+			newReader, err = job.bucket.NewReader(
+				job.cancelCtx,
+				&gcs.ReadObjectRequest{
+					Name:       job.object.Name,
+					Generation: job.object.Generation,
+					Range: &gcs.ByteRange{
+						Start: uint64(start),
+						Limit: uint64(newReaderLimit),
+					},
+					ReadCompressed: job.object.HasContentEncodingGzip(),
+				})
+			if err != nil {
+				err = fmt.Errorf("downloadObjectToFile: error in creating NewReader with start %d and limit %d: %w", start, newReaderLimit, err)
+				return err
+			}
```

Comment:
can you extract this into seperate function so it can be reused in parallel_downloads_job also

Reply:
Could you elaborate what code to put in separate function ?

Assuming you are asking to put {NewReader + copy} logic into a function and then use that in both parallel and non parallel downloads. If yes, then I would avoid that because:
1. NewReader and Copy functions are already abstracted enough and are only called at both the places with their returned errors handled.
2. The read + copy logic is not common in parallel and non parallel downloads. In former case whatever is read is copied to file whereas in latter, a read request of 200MiB is made and data is copied in chunks of 8MiB with updateOffset being called.
3. Even if we go ahead with common function, it's signature will be something like downloadRange(ctx context.Context, dstWriter io.Writer, start int64, end int64, readRequestSize int64, copyChunkSize int64, debugReadType string). This will remove slight duplicacy at the cost of many arguments. 
So, I would prefer keeping it as it is but lmk.

In case you are asking to just abstract out NewReader + err handling code, then also I would prefer keeping it as it is.